Fix tests

romanhabibov opened the pull request.

romanhabibov committed:
test: fix some naming in tests

Fix tests fail appeared after 26e5554.

romanhabibov committed:
Fix bug with wrong urlencoding

Fix bug when the last urlencoded character was ignored. This patch
revert unjustified changes in 26e5554 which reopen #120 and redo
the changes of 1a62080.

Follows up #120

romanhabibov committed:
test: fix test running

Fix test paths when building/running and add test run guide in
README. Remove target "configure", beacuse now it doesn't exist.
Also, disable failing tests.

Totktonada commented on README.md:
```diff
@@ -143,6 +143,24 @@ brew install nginx-full --with-tarantool-module
 
 [Back to contents](#contents)
 
+## Run tests
```
Please, add into the table of contents.

I would also add the 'Back to contents' link as in all other sections.

romanhabibov replied:
Done.

Totktonada commented on README.md:
```diff
@@ -143,6 +143,24 @@ brew install nginx-full --with-tarantool-module
+cp t/ngx_confs/nginx.dev.dyn.conf test-root
+cp t/ngx_confs/tnt_server_test.conf test-root
+cp nginx/conf/mime.types test-root
+tarantool < t/lua.py
```
I guess you meant `t/test.lua`.

romanhabibov replied:
Yes.

Totktonada commented on README.md:
````diff
@@ -143,6 +143,24 @@ brew install nginx-full --with-tarantool-module
+## Run tests
+
+```
+mkdir test-root
+mkdir test-root/logs
+touch test-root/logs/error.log
+touch test-root/logs/access.log
+chmod 777 test-root/logs/error.log
+chmod 777 test-root/logs/access.log
+cp t/ngx_confs/nginx.dev.conf test-root
+cp t/ngx_confs/nginx.dev.dyn.conf test-root
+cp t/ngx_confs/tnt_server_test.conf test-root
+cp nginx/conf/mime.types test-root
+tarantool < t/lua.py
+./nginx/objs/nginx -c nginx.dev.conf
+./t/run_all.sh
````
It looks similar to `t/auto.sh`.

romanhabibov replied:
I updated the patch.

Totktonada commented on src/ngx_http_tnt_module.c:
```diff
@@ -1616,7 +1616,7 @@ ngx_http_tnt_urldecode(ngx_http_request_t *r, ngx_str_t *src)
         c = src->data[s++];
 
-        if (c == '%' && (ngx_uint_t) (s + 2) < src->len) {
+        if (c == '%' && (ngx_uint_t) (s + 2) <= src->len) {
```
It would be good to provide an example of incorrectly parsed URL and add a test case (or mention existing one that reveals the problem).

romanhabibov replied:
Sorry, I pointed the wrong issue. See the new commit message, I just restored old commit with description (in #120) and test.

Totktonada replied:
Now I see. Revertion of one fix was mistakely land inside another fix.

Totktonada commented on src/ngx_http_tnt_version.h:
```diff
@@ -34,6 +34,6 @@
 #define NGX_HTTP_TNT_VERSION_H 1
 
-#define NGX_HTTP_TNT_MODULE_VERSION_STRING "v2.7.2"
+#define NGX_HTTP_TNT_MODULE_VERSION_STRING "v2.7.3"
```
I'm not sure about right release process here, but I find it more clear, when we do it separately from other changes. To be discussed.

romanhabibov replied:
Dropped.

LeonidVas reviewed (changes requested):
Hi! Thank you for the patchset.
See several comments below:
  - In "test/ngx_confs/nginx.dev.conf" `events` have the following configuration: `{worker_connections 4000; multi_accept on;}`. In this case I get the error: "nginx: [emerg] the maximum number of files supported by select() is 1024". The cause of the problem is an attempt to use the `select()` by default. I see two variants for fixes: add `use poll;` or update the value of `worker_connections` to < 1024.
  - Tests "./test/v24_features.py" and "./test/v26_features.py" now fail.
  I think that the failed tests must be commented and corresponding tickets should be created.
  ```
  ┌─[✗]─[leonid@vasya-L460]─[~/work/mail/nginx_upstream_module]
└──╼ ./test/v24_features.py 
[+] Post/Put body
[+] Post body - error
[+] Body sould not pass
[+] Headers out
[+] Unescape issue
Traceback (most recent call last):
  File "/home/leonid/work/mail/nginx_upstream_module/test/http_utils.py", line 223, in get
    res = urllib2.urlopen(req)
  File "/usr/lib/python2.7/urllib2.py", line 154, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib/python2.7/urllib2.py", line 429, in open
    response = self._open(req, data)
  File "/usr/lib/python2.7/urllib2.py", line 447, in _open
    '_open', req)
  File "/usr/lib/python2.7/urllib2.py", line 407, in _call_chain
    result = func(*args)
  File "/usr/lib/python2.7/urllib2.py", line 1235, in http_open
    return self.do_open(httplib.HTTPConnection, req)
  File "/usr/lib/python2.7/urllib2.py", line 1202, in do_open
    h.request(req.get_method(), req.get_selector(), req.data, headers)
  File "/usr/lib/python2.7/httplib.py", line 1103, in request
    self._send_request(method, url, body, headers)
  File "/usr/lib/python2.7/httplib.py", line 1137, in _send_request
    self.putrequest(method, url, **skips)
  File "/usr/lib/python2.7/httplib.py", line 968, in putrequest
    % (url, match.group()))
InvalidURL: URL can't contain control characters. '/unescape?a=some string with spaces' (found at least ' ')

Traceback (most recent call last):
  File "./test/v24_features.py", line 46, in <module>
    result = get_success(preset_method_location, None, {})
  File "/home/leonid/work/mail/nginx_upstream_module/test/http_utils.py", line 259, in get_success
    assert(code == 200), 'expected 200'
AssertionError: expected 200
  ```

  ```
  ┌─[✗]─[leonid@vasya-L460]─[~/work/mail/nginx_upstream_module]
└──╼ ./test/v26_features.py 
[+] basic insert
Traceback (most recent call last):
  File "/home/leonid/work/mail/nginx_upstream_module/test/http_utils.py", line 223, in get
    res = urllib2.urlopen(req)
  File "/usr/lib/python2.7/urllib2.py", line 154, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib/python2.7/urllib2.py", line 429, in open
    response = self._open(req, data)
  File "/usr/lib/python2.7/urllib2.py", line 447, in _open
    '_open', req)
  File "/usr/lib/python2.7/urllib2.py", line 407, in _call_chain
    result = func(*args)
  File "/usr/lib/python2.7/urllib2.py", line 1235, in http_open
    return self.do_open(httplib.HTTPConnection, req)
  File "/usr/lib/python2.7/urllib2.py", line 1202, in do_open
    h.request(req.get_method(), req.get_selector(), req.data, headers)
  File "/usr/lib/python2.7/httplib.py", line 1103, in request
    self._send_request(method, url, body, headers)
  File "/usr/lib/python2.7/httplib.py", line 1137, in _send_request
    self.putrequest(method, url, **skips)
  File "/usr/lib/python2.7/httplib.py", line 968, in putrequest
    % (url, match.group()))
InvalidURL: URL can't contain control characters. '/insert?index=1&string=some big string&float=2.1&double=3.1&bool=True&int=-1000&' (found at least ' ')

Traceback (most recent call last):
  File "./test/v26_features.py", line 34, in <module>
    result = get_success(BASE_URL + '/insert', insert_1, None)
  File "/home/leonid/work/mail/nginx_upstream_module/test/http_utils.py", line 259, in get_success
    assert(code == 200), 'expected 200'
AssertionError: expected 200

  ```

  - When I use `test/auto.sh` it exits with
  ```
  [+] Test - OK
+ for build_type in configure-for-testing configure
+ rm -f nginx/obj/nginx
+ make configure
make: *** No rule to make target 'configure'.  Stop.
  ```
  and it doesn't seem like a good ending. Do you have the same problem?

  - Please use a more detailed description of the problem in commit message for patch "Refix #120". First of all "Refix ..." doesn't seem like a good title. You are not revert the entire commit, you are just fixing some part of it. Why are this changes "unjustified"? Write the reason. Second: "Did I understand correctly from you conversation with Turenko that we already have a test for this change, and test was  failed before your changes?"

  - Use branch naming according to SOP.

romanhabibov committed:
test: reduce number of workers in config

Prevent possible error "nginx: [emerg] the maximum number of files
supported by select() is 1024". The cause of the error is an
attempt to use the select() by default.

romanhabibov commented:
> Hi! Thank you for the patchset.
> See several comments below:
> 
> * In "test/ngx_confs/nginx.dev.conf" `events` have the following configuration: `{worker_connections 4000; multi_accept on;}`. In this case I get the error: "nginx: [emerg] the maximum number of files supported by select() is 1024". The cause of the problem is an attempt to use the `select()` by default. I see two variants for fixes: add `use poll;` or update the value of `worker_connections` to < 1024.

Fixed.

> * Tests "./test/v24_features.py" and "./test/v26_features.py" now fail.

Sorry, I can't fix it now. They don't fail on my Mac. I think the problem is in python's `http_utils` lib version. 

>   I think that the failed tests must be commented and corresponding tickets should be created.
> 
> ```
> ┌─[✗]─[leonid@vasya-L460]─[~/work/mail/nginx_upstream_module]
> └──╼ ./test/v24_features.py 
> [+] Post/Put body
> [+] Post body - error
> [+] Body sould not pass
> [+] Headers out
> [+] Unescape issue
> Traceback (most recent call last):
> File "/home/leonid/work/mail/nginx_upstream_module/test/http_utils.py", line 223, in get
>   res = urllib2.urlopen(req)
> File "/usr/lib/python2.7/urllib2.py", line 154, in urlopen
>   return opener.open(url, data, timeout)
> File "/usr/lib/python2.7/urllib2.py", line 429, in open
>   response = self._open(req, data)
> File "/usr/lib/python2.7/urllib2.py", line 447, in _open
>   '_open', req)
> File "/usr/lib/python2.7/urllib2.py", line 407, in _call_chain
>   result = func(*args)
> File "/usr/lib/python2.7/urllib2.py", line 1235, in http_open
>   return self.do_open(httplib.HTTPConnection, req)
> File "/usr/lib/python2.7/urllib2.py", line 1202, in do_open
>   h.request(req.get_method(), req.get_selector(), req.data, headers)
> File "/usr/lib/python2.7/httplib.py", line 1103, in request
>   self._send_request(method, url, body, headers)
> File "/usr/lib/python2.7/httplib.py", line 1137, in _send_request
>   self.putrequest(method, url, **skips)
> File "/usr/lib/python2.7/httplib.py", line 968, in putrequest
>   % (url, match.group()))
> InvalidURL: URL can't contain control characters. '/unescape?a=some string with spaces' (found at least ' ')
> 
> Traceback (most recent call last):
> File "./test/v24_features.py", line 46, in <module>
>   result = get_success(preset_method_location, None, {})
> File "/home/leonid/work/mail/nginx_upstream_module/test/http_utils.py", line 259, in get_success
>   assert(code == 200), 'expected 200'
> AssertionError: expected 200
> ```
> 
> ```
> ┌─[✗]─[leonid@vasya-L460]─[~/work/mail/nginx_upstream_module]
> └──╼ ./test/v26_features.py 
> [+] basic insert
> Traceback (most recent call last):
> File "/home/leonid/work/mail/nginx_upstream_module/test/http_utils.py", line 223, in get
>   res = urllib2.urlopen(req)
> File "/usr/lib/python2.7/urllib2.py", line 154, in urlopen
>   return opener.open(url, data, timeout)
> File "/usr/lib/python2.7/urllib2.py", line 429, in open
>   response = self._open(req, data)
> File "/usr/lib/python2.7/urllib2.py", line 447, in _open
>   '_open', req)
> File "/usr/lib/python2.7/urllib2.py", line 407, in _call_chain
>   result = func(*args)
> File "/usr/lib/python2.7/urllib2.py", line 1235, in http_open
>   return self.do_open(httplib.HTTPConnection, req)
> File "/usr/lib/python2.7/urllib2.py", line 1202, in do_open
>   h.request(req.get_method(), req.get_selector(), req.data, headers)
> File "/usr/lib/python2.7/httplib.py", line 1103, in request
>   self._send_request(method, url, body, headers)
> File "/usr/lib/python2.7/httplib.py", line 1137, in _send_request
>   self.putrequest(method, url, **skips)
> File "/usr/lib/python2.7/httplib.py", line 968, in putrequest
>   % (url, match.group()))
> InvalidURL: URL can't contain control characters. '/insert?index=1&string=some big string&float=2.1&double=3.1&bool=True&int=-1000&' (found at least ' ')
> 
> Traceback (most recent call last):
> File "./test/v26_features.py", line 34, in <module>
>   result = get_success(BASE_URL + '/insert', insert_1, None)
> File "/home/leonid/work/mail/nginx_upstream_module/test/http_utils.py", line 259, in get_success
>   assert(code == 200), 'expected 200'
> AssertionError: expected 200
> ```
> 
> * When I use `test/auto.sh` it exits with
> 
> ```
> [+] Test - OK
> + for build_type in configure-for-testing configure
> + rm -f nginx/obj/nginx
> + make configure
> make: *** No rule to make target 'configure'.  Stop.
> ```
> 
> and it doesn't seem like a good ending. Do you have the same problem?

Fixed.

> * Please use a more detailed description of the problem in commit message for patch "Refix #120". First of all "Refix ..." doesn't seem like a good title. You are not revert the entire commit, you are just fixing some part of it. Why are this changes "unjustified"? Write the reason.

In fact this changes just revert one line [patch](https://github.com/tarantool/nginx_upstream_module/commit/1a6208051f5a219c7783cfa8910496dcff6faec8) and make the test of this patch failing.

Second: "Did I understand correctly from you conversation with Turenko that we already have a test for this change, and test was  failed before your changes?"
Yes, we have. See [patch](https://github.com/tarantool/nginx_upstream_module/commit/1a6208051f5a219c7783cfa8910496dcff6faec8)
> * Use branch naming according to SOP.

LeonidVas reviewed (approved):
Generally LGTM.
Please update commit message for "Fix #120 again".
The commit title should contain a short description of the patch changes (not "Fix #XXX" or "Add functional for #XXX").
At the end of the commit message add "Follows up #120"

Totktonada reviewed (commented):
That's almost successful for me: I run the testing locally. However several stones were on my way and I would fix / clarify them. Please, consider my comments regarding them.

Totktonada commented on README.md:
```diff
@@ -143,6 +144,23 @@ brew install nginx-full --with-tarantool-module
+itself it creates necessary environment for testing.
+
+For selective running, start:
+  1) nginx in other terminal with `./nginx/objs/nginx -c conf/nginx.conf`,
```
That does not work for me due to several reasons:

1. There is no `conf/nginx.conf` file in the repository root.
2. If `test/ngx_confs/nginx.dev.conf` was assumed, we cannot just add `-c test/ngx_confs/nginx.dev.conf`. The config is searched in a system directory, but even if we'll pass it by a real path, nginx will try to find `mime.types` in a wrong location and will try to write logs into a system directory.
3. All those problems appears, because I tried to follow the readme and assume that the "Build via nginx 'configure'" section is the prerequisite for running the tests. This, however, is not so: here we assume that nginx is built using `make configure-for-testing && make -j` [^1] from this repository root (and `nginx` source repository is placed in the root of this repository). and so `./nginx/objs/nginx` should be called without `-c`. Please, clarify how to build nginx + module in order to run tests: it seems the process differs from building for production use.

Please, test your instructions on a clean repository (without any gitignored files, without nginx sources — from scratch).

[^1]: It is built with custom prefix directory this way (`test-root/`) and nginx configuration for testing is copied inside this directory.

romanhabibov replied:
I've clarified `README.md`. I tested it on clean repository just following README guide. It works fine. I've never used `make configure-for-testing`.

Totktonada replied:
> I've never used `make configure-for-testing`.

And `test/auto.sh`, which calls it?

romanhabibov replied:
No, I used it. Аnyway, I double-checked. Everything works.

Totktonada replied:
I re-read the intruction and, okay, I found where was my mistake. The 'Build from source' instruction suggests `make build-all`, which works for me (together with `make configure-for-testing` afterwards). When I did it the first time, I wrongly followed the "Build via nginx 'configure'" instruction.

romanhabibov committed:
test: get VERBOSE from environment

Get debugging flag VERBOSE from environment if it exists.

romanhabibov committed:
test: keep logs in files

Log output from python test files to coresponding log files.

romanhabibov committed:
test: replace python shebangs

Replace python shebangs from `#!/usr/bin/python` to
`#!/usr/bin/env python` in order to run the tests using a python
interpreter from a virtualenv.

romanhabibov committed:
test: add recent nginx version

Add recent nginx version 1.19.7 to test the module with it.

romanhabibov commented on test/auto.sh:
```diff
@@ -0,0 +1,34 @@
+  nginx_pid=$!
+  tarantool $WORK_DIR/test.lua 2> /dev/null &
+  tnt_pid=$!
+  ./$WORK_DIR/run_all.sh
```
Done.

romanhabibov commented on test/v25_features.py:
```diff
@@ -28,7 +28,7 @@
 assert out['result'][0]["headers"]["Content-Type"] == \
         default_headers["Content-Type"], "Content-Type not equals"
-assert out['result'][0]['body'][0] == {"a": "b"}, "not expected result"
+assert out['result'][0]['args_urlencoded'][0] == {"a": "b"}, "not expected result"
```
Replaced.

Totktonada replied:
Thanks!

romanhabibov commented on test/ngx_confs/nginx.dev.conf:
```diff
@@ -10,6 +10,6 @@ pid logs/nginx.pid;
 error_log stderr;
 
-events {worker_connections 4000; multi_accept on;}
+events {worker_connections 1023; multi_accept on;}
```
Done.

romanhabibov commented on README.md:
```diff
@@ -143,6 +144,23 @@ brew install nginx-full --with-tarantool-module
 
+## Test run and notes for contributors
+
+Firstly, run all tests with `./test/auto.sh`. Before test running
```
Disabled, because this pull request has gone too far...

Totktonada replied:
Okay.

romanhabibov commented on README.md:
```diff
@@ -143,6 +144,23 @@ brew install nginx-full --with-tarantool-module
+## Test run and notes for contributors
+
+Firstly, run all tests with `./test/auto.sh`. Before test running
+itself it creates necessary environment for testing.
```
Clarified.

Totktonada replied:
Thanks!

romanhabibov commented on README.md:
```diff
@@ -143,6 +144,23 @@ brew install nginx-full --with-tarantool-module
+If you want to add new test, see examples: `*_features.py` files.
+Don't forget to include it to `test/run_all.sh`.
+
+You can use `VERBOSE` flag in `test/http_utils.py` for debugging.
```
Done.

Totktonada replied:
Nice, thanks!

romanhabibov commented on README.md:
```diff
@@ -143,6 +144,23 @@ brew install nginx-full --with-tarantool-module
+
+For selective running, start:
+  1) nginx in other terminal with `./nginx/objs/nginx -c conf/nginx.conf`,
+  2) Tarantool in another terminal with `tarantool < test/test.lua`,
```
Done.

romanhabibov commented on test/auto.sh:
```diff
@@ -0,0 +1,34 @@
+  git checkout $ver_tag
+  cd -
+
+  rm -f nginx/obj/nginx
```
Fixed.

romanhabibov commented on test/auto.sh:
```diff
@@ -0,0 +1,34 @@
+set -x -e
+WORK_DIR=test
+
+for ver_tag in `cat $WORK_DIR/ngx_versions_list`; do
```
Added.

romanhabibov replied:
Done.

Totktonada replied:
Now 1.19.7 is out. I would be nice to have it in testing.

Totktonada replied:
Updated.

Totktonada commented on README.md:
```diff
@@ -141,6 +144,28 @@ brew install nginx-full --with-tarantool-module
+
+## Test run and notes for contributors
+
+This paragraph is actually only if you chose [build from source](#build-from-source)
```
Grammar:

1. 'Actually' is adverb, 'actual' is adjective. Use the latter here.
2. 'Chose' is the past tense form, let's use infinitive form 'choose' here.

romanhabibov replied:
Fixed.

Totktonada commented on test/run_all.sh:
```diff
@@ -4,76 +4,35 @@ set -e #-x
-  clients_pids="$clients_pids $!"
+  echo "[+] try: $i"
+  for test_file in "${test_files[@]}"; do
+    `$WORK_DIR/$test_file.py 1> $LOG_PATH/$test_file.log || (
```
It'll rewrite logs from non-parallel run. I would name them differently.

BTW, you appending (`>>`) logs for non-parallel runs to existing logs, but replacing (`>`) old logs for parallel run. There should be some consistency.

romanhabibov replied:
Done.

romanhabibov replied:
I choose `>`.

Totktonada replied:
Parallel runs of one test still write to the same log file. It should contain an iteration number.

romanhabibov replied:
Done.

Totktonada commented on test/run_all.sh:
```diff
@@ -4,76 +4,35 @@ set -e #-x
 
 WORK_DIR=test
 NGINX_PREFIX=$PWD/test-root
+LOG_PATH=$NGINX_PREFIX/logs
```
I would also this path to README or show log file paths on the terminal during testing. The latter looks better.

Otherwise one will need to dive into the code to find logs location.

romanhabibov replied:
Done.

Totktonada replied:
The log directory will be shown only after successful execution of all tests. If any of the tests fails, it will not be shown. I would show a log file path before each test run or show a directory with logs before executing a first test.

romanhabibov replied:
I prefer to show a log file path before each test.

Totktonada replied:
I see that you do it before all tests, but it is okay anyway.

Totktonada reviewed (approved):
I have no objections anymore.

The test are passed locally.

Totktonada commented:
Chug-chug!
